src/app/app-platform/ContentAppPlatform.cpp

```diff
@@ -108,38 +110,102 @@ EndpointId ContentAppPlatform::AddContentApp(ContentApp * app, EmberAfEndpointTy
-                {
-                    mCurrentEndpointId = mFirstDynamicEndpointId;
-                }
+                ChipLogProgress(DeviceLayer, "Added ContentApp %s to dynamic endpoint %d (index=%d)", vendorApp.applicationId,
+                                mCurrentEndpointId, index);
+                app->SetEndpointId(mCurrentEndpointId);
+                IncrementCurrentEndpointID();
+                return app->GetEndpointId();
             }
+            else if (ret != EMBER_ZCL_STATUS_DUPLICATE_EXISTS)
+            {
+                ChipLogError(DeviceLayer, "Adding ContentApp error=%d", ret);
+                return kNoCurrentEndpointId;
+            }
+            IncrementCurrentEndpointID();
+        } while (initEndpointId != mCurrentEndpointId);
+        ChipLogError(DeviceLayer, "Failed to add dynamic endpoint: No endpoints available!");
+        return kNoCurrentEndpointId;
+    }
+    ChipLogError(DeviceLayer, "Failed to add dynamic endpoint: max endpoint count reached!");
+    return kNoCurrentEndpointId;
+}
+
+EndpointId ContentAppPlatform::AddContentApp(ContentApp * app, EmberAfEndpointType * ep,
+                                             const Span<DataVersion> & dataVersionStorage,
+                                             const Span<const EmberAfDeviceType> & deviceTypeList, EndpointId desiredEndpointId)
+{
+    CatalogVendorApp vendorApp = app->GetApplicationBasicDelegate()->GetCatalogVendorApp();
+
+    ChipLogProgress(DeviceLayer, "Adding ContentApp with appid %s ", vendorApp.applicationId);
+    uint8_t index = 0;
+    // check if already loaded
+    while (index < CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT)
+    {
+        if (mContentApps[index] == app)
+        {
+            ChipLogProgress(DeviceLayer, "Already added");
+            // already added, return endpointId of already added endpoint.
+            // desired endpointId does not have any impact
+            return app->GetEndpointId();
         }
         index++;
     }
-    ChipLogError(DeviceLayer, "Failed to add dynamic endpoint: No endpoints available!");
+
+    if (desiredEndpointId < FIXED_ENDPOINT_COUNT ||
+        emberAfGetDynamicIndexFromEndpoint(desiredEndpointId) != kEmberInvalidEndpointIndex)
+    {
+        // invalid desiredEndpointId
+        ChipLogError(DeviceLayer, "Failed to add dynamic endpoint: desired endpointID is invalid!");
+        return kNoCurrentEndpointId;
+    }
+
+    index = 0;
+    while (index < CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT)
+    {
+        if (mContentApps[index] != nullptr)
+        {
+            index++;
+            continue;
+        }
+        mContentApps[index] = app;
+        EmberAfStatus ret   = emberAfSetDynamicEndpoint(index, desiredEndpointId, ep, dataVersionStorage, deviceTypeList);
+        if (ret != EMBER_ZCL_STATUS_SUCCESS)
+        {
+            ChipLogError(DeviceLayer, "Adding ContentApp error=%d", ret);
```

Comment:
same question here: do we have to undo any operations when we return failure?

Since `this` is stateful and the state is modified, it seems that the error should undo state changes. Otherwise returns on line 173 and on line 189 have the same return code but different sideffects.

Reply:
Moved setting the entry in the content apps array only to the success case.